File: Queue/LL_Queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LL_Queue:
    def __init__(self):
        self.front = self.rear = None

    def enqueue(self, data):
        new_node = Node(data)
        if self.rear is None: # This is the firsts element in the queue 
            self.front = self.rear = new_node
            print(f"Node {data} enqueued")
            return
        self.rear.next = new_node
        self.rear = new_node
        print(f"Node {data} enqueued")

    def dequeue(self):
        if self.isEmpty():
            print("Queue is empty!")
            return
        dequeued_value = self.front.data
        self.front = self.front.next
        if self.front is None: #After deletion no node i.e deleted node is the only node in the queue
            self.rear = None  # Set rear too null as no element in the queue
        return dequeued_value

    def display(self):
        if self.isEmpty():
            print("Queue is empty!")
            return
        print("Queue element: ",end="")
        temp = self.front
        while temp:
            print(temp.data, end=" ")
            temp = temp.next
        print()
        
    def peek(self):
        return self.front.data;
        
    def lastElement(self):
        return self.rear.data;
        
    def isEmpty(self):
        return self.front == None 


queue = LL_Queue()
queue.enqueue(10)
queue.enqueue(20)
queue.enqueue(30)
queue.display()

print(f"Dequeued element: {queue.dequeue()}")
queue.display()

print(f"First element: {queue.peek()}")
print(f"Last element: {queue.lastElement()}")

print(f"Dequeued element: {queue.dequeue()}")
print(f"Dequeued element: {queue.dequeue()}")
print(f"Dequeued element: {queue.dequeue()}")

print(f"Is queue empty?: {queue.isEmpty()}")

File: Queue/LL_Queue.java
class Node{
    int data;
    Node next;
    
    Node(int data)
    {
        this.data = data;
        this.next = null;
    }
}

class Main{
    
    private Node rear, front;
    
    public Main()
    {
        rear = front = null;
    }
    
    public void enqueue(int data)
    {
        Node nn = new Node(data);
        if(rear == null)
        {
            front = rear = nn;
            System.out.println("Node "+data+" is enqueued to the queue");
            return;
        }
        rear.next = nn;
        rear = nn;
        System.out.println("Node "+data+" is enqueued to the queue");
    }
    
    public int dequeue()
    {
        if(front == null)
        {
            System.out.println("Queue is empty");
            return -1;
        }
        
        int deq_ele = front.data;
        front = front.next; 
        
        if(rear == null) front = null;
        
        return deq_ele;
    }
    
    public void display()
    {
        if(front == null)
        {
            System.out.println("Queue is empty");
            return;
        }
        Node temp = front;
        System.out.print("Queue elements: ");
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }
    
    
    public boolean isEmpty()
    {
        return front == null;
    }
    
    public int peek()
    {
        return front.data;
    }
    
    public int lastAppendedData()
    {
        return rear.data;
    }
    
    public static void main(String args[])
    {
        Main ll = new Main();
        ll.enqueue(10);
        ll.enqueue(20);
        ll.enqueue(30);
        ll.enqueue(40);
        ll.enqueue(50);
        
        ll.display();
        
        System.out.println("Dequeued element is: "+ll.dequeue());
        ll.display();
        
        
        System.out.println("Queue is empty: "+ll.isEmpty());
        System.out.println("Queue peek(next element to the removed): "+ll.peek());
        System.out.println("Queue last inserted element: "+ll.lastAppendedData());
    }
    
}